File: 4-Behavioral Patterns/6-Command Pattern/Restaurant Example/Python/customer.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Customer module that actually uses Command Pattern.
"""

__author__ = 'Ziang Lu'

from restaurant import CleanCommand, Cleaner, Cook, Waiter


def main():
    restaurant_cook = Cook()  # Receiver
    restaurant_cleaner = Cleaner()  # Receiver
    restaurant_waiter = Waiter(restaurant_cook)  # Invoker

    # Request the waiter [Invoker] to place an order [Command]
    restaurant_waiter.place_order('Kung Pao Chicken', 'Sesame Chicken')

    print()

    # Create a clean command [Command] by passing the cleaner [Receiver]
    clean_command = CleanCommand(restaurant_cleaner)
    # Let the waiter [Invoker] to execute the clean command [Command]
    restaurant_waiter.clean(clean_command)


if __name__ == '__main__':
    main()

# Output:
# Waiter [Invoker] has received an order with items ('Kung Pao Chicken', 'Sesame Chicken')
# A new order [Command] has been created.
# Kung Pao Chicken has been added to the order.
# Sesame Chicken has been added to the order.
# Waiter [Invoker] starts executing the command...
# Cook [Receiver] is cooking Sesame Chicken.
# Cook [Receiver] is cooking Kung Pao Chicken.
#
# A new clean command [Command] has been created.
# Waiter [Invoker] has received the clean command [Command] and starts executing the command...
# Cleaner [Receiver] is cleaning up the table.

File: 4-Behavioral Patterns/6-Command Pattern/Restaurant Example/Java/runner/Customer.java
package runner;

import restaurant.CleanCommand;
import restaurant.Cleaner;
import restaurant.Cook;
import restaurant.Waiter;

/**
 * Application that actually uses Command Pattern.
 *
 * @author Ziang Lu
 */
public class Customer {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        Cook restaurantCook = new Cook(); // Receiver
        Cleaner restaurantCleaner = new Cleaner(); // Receiver
        Waiter restaurantWaiter = new Waiter(restaurantCook); // Invoker

        // Request the waiter [Invoker] to place an order [Command]
        restaurantWaiter.placeOrder("Kung Pao Chicken", "Sesame Chicken");

        System.out.println();

        // Create a clean command [Command] by passing the cleaner [Receiver]
        CleanCommand cleanCommand = new CleanCommand(restaurantCleaner);
        // Let the waiter [Invoker] to execute the clean command [Command]
        restaurantWaiter.clean(cleanCommand);

        /*
         * Output:
         * Waiter [Invoker] has received an order with items: [Kung Pao Chicken, Sesame Chicken].
         * A new order [Command] has been created.
         * Kung Pao Chicken has been added to the order.
         * Sesame Chicken has been added to the order.
         * Waiter [Invoker] starts executing the command...
         * Cook [Receiver] is cooking Sesame Chicken.
         * Cook [Receiver] is cooking Kung Pao Chicken.
         *
         * A new clean command [Command] has been created.
         * Waiter [Invoker] has received the clean command [Command] starts executing the command...
         * Cleaner [Receiver] is cleaning up the table.
         */
    }

}
